ui-v2/app/templates/components/discovery-chain.hbs

```diff
@@ -0,0 +1,126 @@
+<style>
+  {{#if selected.nodes }}
+    {{selected.nodes}} {
+      opacity: 1 !important;
+
+      background-color: {{css-var '--white'}};
+      border-color: {{css-var '--gray-500'}};
+    }
+  {{/if}}
+  {{#if selected.edges }}
+    {{selected.edges}} {
+      opacity: 1;
+    }
+  {{/if}}
+</style>
```

Comment:
This is an anti-pattern in the sense that I would never think to look at a component's templates to find CSS.

Why not use `active` classes or something similar instead of rewriting the underlying stylesheet?

Reply:
Yeah so this is obviously the tough one with this entire feature. I can't quite remember the exact detail - but, all the data here comes in as keyed objects, not arrays. The keys are arbitrary id's so you can't specify them in computed properties (I think I mentioned this problem offline a while ago).

This approach is so we can highlight the related nodes/edges in the graph.

This means you have to detect a change on the root of the entire chain. So if you change a property you probably have to re-filter out the entire chain with loops and conditionals again, recheck all the dom positions etc etc. Basically lots of stuff.

Once it's done that, it will have to mutate multiple (possibly lots) of DOM nodes in order to apply an 'active' class.

Using this approach we just use our graphing lib to give us the id's of all the things that should be highlighted (as we would have to do with the above approach anyway), make one single DOM mutation, and let CSS/the browser do the rest. Seems to me to be the simplest, least intensive approach. There might be other ways to do it, but this works great and is simple so I went with it. I started with this DOM poking version of this, but figured I could use ember templates to do it instead to make it as ember-y as possible.

Agree with the 'never thinking to look in the template for styles though' I'll add a comment in the stylesheet so its clear there's a little bit of CSS in the template. I'll drop it in https://github.com/hashicorp/consul/pull/6939

Reply:
I don't know the numbers off the top of my head, but my (potentially very wrong) intuition is that even if going the route of having an array representation of the underlying data will mean decorating N DOM nodes with a class marker, the alternative of injecting a stylesheet, re-baking selectors, and verifying the entire dom tree is slower.

It's similar work to what you're avoiding, just done by the browser instead. Granted it could be highly optimized C++ instructions, so I can't say for sure if there's actually a performance tradeoff here.

Reply:
> Granted it could be highly optimized C++ instructions

Sounds like a good assumption that would be faster 😀  I could be completely wrong of course, but I don't think it matters that much here.

I think the thing that matters more is - the best code is zero code. You can't have bugs in code if it doesn't exist, and it means I can spend time doing other stuff (better performance 😄 )